### PR TITLE
CPM-732: Add Controller to get identifier attributes

### PR DESCRIPTION
**Pull request by pierallard**

### CODE REVIEW

File: components/identifier-generator/back/src/Infrastructure/Controller/GetIdentifierAttributesController.php [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+use Akeneo\UserManagement\Bundle\Context\UserContext;
+use Symfony\Component\HttpFoundation\JsonResponse;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpKernel\Exception\AccessDeniedHttpException;
+
+final class GetIdentifierAttributesController
+{
+    // The limit of the number of identifier attributes in the PIM
```
Comment (pierallard) [inline components/identifier-generator/back/src/Infrastructure/Controller/GetIdentifierAttributesController.php:17]:
```suggestion
    // The maximum count of identifier attributes in the PIM
```

File: components/identifier-generator/back/src/Infrastructure/Controller/GetIdentifierAttributesController.php [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+            return new RedirectResponse('/');
+        }
+
+        if (!$this->securityFacade->isGranted('pim_enrich_attribute_index')) {
+            throw new AccessDeniedHttpException('Access forbidden. You are not allowed to list attributes.');
+        }
+
+        $attributes = $this->findFlattenAttributes->execute(
+            $this->userContext->getCurrentLocaleCode(),
+        self::MAX_RESULTS,
```
Comment (fchoquet) [inline components/identifier-generator/back/src/Infrastructure/Controller/GetIdentifierAttributesController.php:38]: check indentation here

File: components/identifier-generator/back/src/Infrastructure/Controller/GetIdentifierAttributesController.php [hunk 3]
```diff
@@ -0,0 +1,54 @@
+<?php
+
```
Comment (fchoquet) [inline components/identifier-generator/back/src/Infrastructure/Controller/GetIdentifierAttributesController.php:2]: Any reason not to use strict_types, or is it an omission?

File: components/identifier-generator/back/src/Infrastructure/Controller/GetIdentifierAttributesController.php [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+use Symfony\Component\HttpFoundation\JsonResponse;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpKernel\Exception\AccessDeniedHttpException;
+
+final class GetIdentifierAttributesController
+{
+    // The maximum count of identifier attributes in the PIM
+    const MAX_RESULTS = 1;
```
Comment (fchoquet) [inline components/identifier-generator/back/src/Infrastructure/Controller/GetIdentifierAttributesController.php:18]: If not meant to be accessed externally:

```suggestion
    private const MAX_RESULTS = 1;
```

**Review by fchoquet — COMMENTED** [review]
Looks good! A few comments / questions

File: components/identifier-generator/back/src/Infrastructure/Controller/GetIdentifierAttributesController.php (hunk 3 above)
Comment (cemorin) [reply components/identifier-generator/back/src/Infrastructure/Controller/GetIdentifierAttributesController.php:2]: it was an omission :)

**Review by Beninho — APPROVED** [review]

File: components/identifier-generator/Makefile [hunk 5]
```diff
@@ -0,0 +1,18 @@
+IDENTIFIER_GENERATOR_PATH ?= components/identifier-generator
+IDENTIFIER_GENERATOR_BOOTSTRAP_PATH ?= /srv/pim/config/bootstrap.php
+IDENTIFIER_GENERATOR_YARN = $(YARN_RUN)
+
+# TESTS BACK
+.PHONY: identifier-generator-integration-back
```
Comment (pierallard) [inline components/identifier-generator/Makefile:6]: 1) I'm not seeing where do you call these tests on CI.
2) Why not put the makefile in the same place than the other makefiles ?

File: components/identifier-generator/Makefile [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+identifier-generator-integration-back:
+	APP_ENV=test $(PHP_RUN) vendor/bin/phpunit \
+		--configuration $(IDENTIFIER_GENERATOR_PATH)/back/ \
+		--bootstrap $(IDENTIFIER_GENERATOR_BOOTSTRAP_PATH) \
+		--testsuite Identifier_Generator_Integration \
+		--order-by random \
+		--log-junit var/tests/phpunit/phpunit_identifier_generator_integration.xml ${O}
+
+# TESTS FRONT
```
Comment (pierallard) [inline components/identifier-generator/Makefile:15]: PLEASE AVOID CAPS LOCK THANKS

File: components/identifier-generator/back/src/Infrastructure/Symfony/Resources/translations/jsmessages.en_US.yml [hunk 7]
```diff
@@ -1,2 +1,5 @@
 pim_title:
     akeneo_identifier_generator_index: "Identifier generators"
+pim_error:
+    unauthorized: "You can not list attributes. Please contact your administrator to change your rights."
```
Comment (pierallard) [inline components/identifier-generator/back/src/Infrastructure/Symfony/Resources/translations/jsmessages.en_US.yml:4]:
```suggestion
    unauthorized: "You cannot list attributes. Please contact your administrator to change your rights."
```

File: components/identifier-generator/back/tests/EndToEnd/Infrastructure/Controller/GetIdentifierAttributesControllerIntegration.php [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        $response = $this->client->getResponse();
+        Assert::assertSame(Response::HTTP_FORBIDDEN, $response->getStatusCode());
+    }
+
+    protected function getConfiguration(): Configuration
+    {
+        return $this->catalog->useTechnicalCatalog();
+    }
+
+    private function getWebClientHelper(): WebClientHelper {
```
Comment (pierallard) [inline components/identifier-generator/back/tests/EndToEnd/Infrastructure/Controller/GetIdentifierAttributesControllerIntegration.php:73]:
```suggestion
    private function getWebClientHelper(): WebClientHelper
    {
```

File: components/identifier-generator/front/src/feature/IdentifierGeneratorApp.tsx [hunk 9]
```diff
@@ -1,41 +1,44 @@
 import React from 'react';
 import {Breadcrumb, Button, Helper} from 'akeneo-design-system';
 import {PageHeader, PimView, useTranslate} from '@akeneo-pim-community/shared';
+import {IdentifierAttributeSelector} from './components';
+import {QueryClient, QueryClientProvider} from 'react-query';
+
+// Create a client
```
Comment (pierallard) [inline components/identifier-generator/front/src/feature/IdentifierGeneratorApp.tsx:7]: Remove this comment

File: components/identifier-generator/front/src/feature/models/attributes.d.ts [hunk 10]
```diff
@@ -0,0 +1,6 @@
+type Attribute = {
```
Comment (pierallard) [inline components/identifier-generator/front/src/feature/models/attributes.d.ts:1]: Maybe "FlattenAttribute", as we probably need other attributes in the future

File: components/identifier-generator/back/tests/EndToEnd/Infrastructure/Controller/GetIdentifierAttributesControllerIntegration.php [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+            'akeneo_identifier_generator_get_identifier_attributes',
+            [],
+            'GET',
+            [
+                'HTTP_X-Requested-With' => 'XMLHttpRequest'
+            ]
+        );
+        $response = $this->client->getResponse();
+        Assert::assertSame(Response::HTTP_OK, $response->getStatusCode());
+        Assert::assertSame('[{"code":"coucou","label":"[sku]"}]', $response->getContent());
```
Comment (mmetayer) [inline components/identifier-generator/back/tests/EndToEnd/Infrastructure/Controller/GetIdentifierAttributesControllerIntegration.php:47]: coucou?

File: components/identifier-generator/back/tests/EndToEnd/Infrastructure/Controller/GetIdentifierAttributesControllerIntegration.php [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+        $this->getWebClientHelper()->callRoute(
+            $this->client,
+            'akeneo_identifier_generator_get_identifier_attributes',
+            [],
+            'GET',
+            [
+                'HTTP_X-Requested-With' => 'XMLHttpRequest'
+            ]
+        );
+        $response = $this->client->getResponse();
```
Comment (mmetayer) [inline components/identifier-generator/back/tests/EndToEnd/Infrastructure/Controller/GetIdentifierAttributesControllerIntegration.php:45]: would probably be more convenient (and readable) to have a protected `loginAs($username)` and `callRoute()` $method in the abstract class

File: components/identifier-generator/back/src/Infrastructure/Controller/GetIdentifierAttributesController.php [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpKernel\Exception\AccessDeniedHttpException;
+
+final class GetIdentifierAttributesController
+{
+    // The maximum count of identifier attributes in the PIM
+    private const MAX_RESULTS = 1;
+
+    public function __construct(
+        private FindFlattenAttributesInterface $findFlattenAttributes,
```
Comment (mmetayer) [inline components/identifier-generator/back/src/Infrastructure/Controller/GetIdentifierAttributesController.php:23]: BTW, IMO we should implement our own query

File: components/identifier-generator/back/src/Infrastructure/Controller/GetIdentifierAttributesController.php [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    public function __invoke(Request $request): Response {
+        if (!$request->isXmlHttpRequest()) {
+            return new RedirectResponse('/');
+        }
+
+        if (!$this->securityFacade->isGranted('pim_enrich_attribute_index')) {
+            throw new AccessDeniedHttpException('Access forbidden. You are not allowed to list attributes.');
+        }
```
[the remark this answers is not in the document]
Comment (cemorin) [reply components/identifier-generator/back/src/Infrastructure/Controller/GetIdentifierAttributesController.php:36]: Remove, and create a card to check with Christophe if we have to check list.Attributes
And create card to talk about our own permissions

**Comment by cemorin** [comment]
split the PR front/back

**Pull request closed without merging**